README.md

```diff
@@ -1,2 +1,176 @@
+* Issuance
+  1. The key is hashed using BCrypt (variable) and used as `<signature>`.
+  2. The client is presented with `<key>.<signature>`.
+  3. The signature is encrypted and stored in the database using AES (variable).
+* Validation
+ 1. The client presents `<key>.<signature>`.
+ 2. It is validated if <key> matches <signature> using BCrypt (variable).
+ 3. The signature is encrypted using AES (variable).
+ 4. The encrypted signature is looked up in the database, failing validation if no such row exists.
+ 5. They key is considered valid and is now subject for other validations, like audience, redirect or state matching.
+
+#### Encrypt credentials at rest
+
+Credentials (tokens, passwords and secrets) are always encrypted at rest.
+
+#### Implement peer reviewed IETF Standards
+
+Fosite implements [rfc6749](https://tools.ietf.org/html/rfc6749) and enforces countermeasures suggested in [rfc6819](https://tools.ietf.org/html/rfc6819).
+
+### Provide extensibility and interoperability
+
+... because OAuth2 is an extensible and flexible **framework**. Fosite let's you register new response types, new grant
+types and new response key value pares. This is useful, if you want to provide OpenID Connect on top of your
+OAuth2 stack. Or custom assertions, what ever you like and as long as it is secure. ;)
+
+### Tokens
+
+Tokens are formatted as `<key>.<signature>`. This is beneficial if you want to keep tokens encrypted at rest.
+To validate a token in a OAuth2 grant, you could first check if the key matches the signature and then lookup the signature
+in your persistent storage (e.g. MySQL). If your persistent storage is intruded (e.g. by SQL injection), an attacker would
+only have access to the token signatures and would be, because he does not know the key, unable to use them for authorization.
+
+A token generated by `generator.CryptoGenerator` looks like:
+
+```
+GUULhK6Od/7UAlnKvMau8APHSKXSRwm9aoOk56SHBns.JDJhJDEwJDdwVmpCQmJLYzM2VDg1VHJ5aEdVOE81NVdRSkt6bHBHR1QwOC9pbTNFWmpQRXliTWRPeDQy
+```
+
+## Usage
+
+This section is WIP and we welcome discussions via PRs or in the issues.
+
+### Store
+
+### Authorize Endpoint
+
+```go
+var r *http.Request // we're assuming that we are inside a http.Handler
+var rw http.ResponseWriter  // we're assuming that we are inside a http.Handler
+
+var store fosite.Storage // needs to be implemented or by using some library
+config := fosite.NewDefaultConfig()
+oauth := fosite.NewOAuth(config)
+authorizeRequest, err := oauth.NewAuthorizeRequest(r, store)
+if err != nil {
+    oauth.RedirectError(rw, error)
+    // or, for example: oauth.WriteError(rw, error)
+    return
+}
+
+// you have now access to authorizeRequest.Scope, ...Code ...ResponseTypes ...Scopes ...
+
+// decide what to do based on scope and response type
+// e.g: response, err = oauth.HandleAuthorizeRequest(authorizeRequest)
+
+// set up a session
+// session := oauth2.NewAuthorizeSession(123)
```

Comment:
Is this going to use the JSONSession? I don't totally understand the use-case for this yet. Otherwise I like the separation of concerns. is the fosite.NewOAuth(config) necessary or can it requests just come directly from the config -- similar to the oauth2 client.